### PR TITLE
provider/do: add support for Digital Ocean

### PR DESCRIPTION
**Pull request by rjeczalik**

### CODE REVIEW

**Review by cihangir — APPROVED** [review]

**Comment by cihangir** [comment]
🎉

File: go/src/koding/kites/kloud/provider/do/schema.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		return errors.New("key fingerprint cannot be empty")
+	}
+
+	return nil
+}
+
+// isValid checks whether the given region is valid or not
+func (r Region) isValid() bool {
+	for _, validRegion := range validRegions {
+		if validRegion == string(r) {
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/do/schema.go]: Not a bug but `var validRegion = []Region{ /* .... */ }` will drop this type casting.

File: go/src/koding/kites/kloud/provider/do/schema_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
+package do
+
+import "testing"
+
+func TestCredential_Valid(t *testing.T) {
+	cred := &Credential{
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/do/schema_test.go]: Another (small) note: all tests in this file could be merged into one with table style fashion.

File: go/src/koding/kites/kloud/provider/do/stack.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+	// we might have multipel droplets, iterate over all of them
+	for dropletName, droplet := range resource.Droplet {
+		// Do not overwrite SSH key pair with the bootstrap one
+		// when user sets it explicitly in a template.
+		if s, ok := droplet["ssh_keys"]; !ok {
+			droplet["ssh_keys"] = []int{keyID}
+		} else if keyIds, ok := s.([]int); ok {
+			keys := []int{keyID}
+			if len(keyIds) != 0 {
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/do/stack.go]: even `var keyIds []int = nil` can be expanded `keyIds...` so no need for this check.

File: go/src/koding/kites/kloud/provider/do/stack.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+		// kite id.
+		count := 1
+		if n, ok := droplet["count"].(int); ok && n > 1 {
+			count = n
+		}
+
+		labels := []string{dropletName}
+		if count > 1 {
+			for i := 0; i < count; i++ {
+				labels = append(labels, fmt.Sprintf("%s.%d", dropletName, i))
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/do/stack.go]: hmm if `count = 2` then we will have `labels = []string{"droplet", "droplet.0", "droplet.1"}`. Doesn't seem right.

**Review by ppknap — COMMENTED** [review]
Mostly small issues :)

File: go/src/koding/kites/kloud/provider/do/machine.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+			if err != nil {
+				return err
+			}
+			if action.Status == actionErrored {
+				return errors.New(action.String())
+			}
+			if action.CompletedAt != nil || action.Status == actionCompleted {
+				return nil
+			}
+		case <-ctx.Done():
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/do/machine.go]: No we don't, moreover context are used wrongly - for storing global-scoped data. I'm in ongoing effort to have that all fixed.

**Comment by rjeczalik** [comment]
> Mostly small issues
@ppknap All valid, will address them in separate PR after https://github.com/koding/koding/pull/9182.

File: go/src/koding/kites/kloud/provider/do/machine_test.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+		Machine:    &models.Machine{},
+		Session:    nil,
+		Credential: doProvider.Schema.NewCredential(),
+		Bootstrap:  doProvider.Schema.NewBootstrap(),
+		Metadata: doProvider.Schema.NewMetadata(&stack.Machine{
+			Attributes: map[string]string{
+				"droplet_id": strconv.Itoa(dropletID),
+			},
+		}),
+		Provider: "do",
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/do/machine_test.go]: i was not sure if this field is used to create terraform template where they use `digitalocean` as a prefix [link](https://www.terraform.io/docs/providers/do/r/droplet.html).

**Pull request merged**